iris_ugrid/ugrid_cf_reader.py

```diff
@@ -0,0 +1,240 @@
+                nodes_dim_name = nodes_var.dimensions[0]
+                meshdims_map[nodes_dim_name] = (mesh, "node")
+        self.meshdims_map = meshdims_map
+
+        # Create a CFReader object which skips the UGRID-related variables.
+        kwargs["exclude_var_names"] = exclude_vars
+        self.cfreader = iris.fileformats.cf.CFReader(
+            self.dataset, *args, **kwargs
+        )
+
+    def complete_unstructured_cube(self, cube):
+        """
+        Cube post-processing hook called by :func:`iris.fileformats.netcdf.load_cubes`, to add the unstructured info
+        onto a cube newly created by the `self.cfreader`.
+
+        Add the ".ugrid" property to a cube loaded with the `self.cfreader`.
+        We identify the unstructured-grid dimension of the cube (if any), and
+        attach a suitable CubeUgrid object, linking the cube mesh dimension to
+        an element-type (aka "mesh_location") of a mesh.
+        """
+        # Set a 'cube.ugrid' property.
+        data_var = self.dataset.variables[cube.var_name]
+        meshes_info = [
+            (i_dim, self.meshdims_map.get(dim_name))
+            for i_dim, dim_name in enumerate(data_var.dimensions)
+            if dim_name in self.meshdims_map
+        ]
+        if len(meshes_info) > 1:
+            msg = "Cube maps more than one mesh dimension: {}"
+            raise ValueError(msg.format(meshes_info))
+        if meshes_info:
+            i_dim, (mesh, mesh_location) = meshes_info[0]
+            mesh_var = self.dataset.variables[mesh.mesh_name]
+
+            topology_dimension = mesh_var.getncattr("topology_dimension")
+            node_coordinates = []
+            for node_var_name in mesh_var.getncattr("node_coordinates").split(
+                " "
+            ):
+                node_var = self.dataset.variables[node_var_name]
+                name = (
+                    getattr(node_var, "standard_name", None)
+                    or getattr(node_var, "long_name", None)
+                    or node_var_name
+                )
+                node_coordinates.append(name)
+
+            cube.ugrid = CubeUgrid(
+                cube_dim=i_dim,
+                grid=mesh,
+                mesh_location=mesh_location,
+                topology_dimension=topology_dimension,
+                node_coordinates=sorted(node_coordinates),
+            )
+        else:
+            # Add an empty 'cube.ugrid' to all cubes otherwise.
+            cube.ugrid = None
+
+
+def load_cubes(filenames, callback=None):
+    def create_ugrid_reader(filename):
+        """Function to create a CFReader object which is patched to do ugrid operations."""
+        ugrid_reader = UGridCFReader(filename)
+        # We must return the inner, actual CFReader object..
+        inner_cf_reader = ugrid_reader.cfreader
+        # .. but we also create on it a 'cube_completion' method, which the loader operation will call for us.
+        inner_cf_reader.cube_completion_adjust = (
```

Comment:
Also discussed offline with @trexfeathers 
It now seems to me (who originally did it this way) that here also it would be cleaner to use inheritance, rather than attaching an extra function property.  Though in this case the generality of the mechanism seem to be OK.
Fixing this would require a further change to Iris, but I think we will need one anyway to enable CF loading to produce "Ucube"s in place of Cubes.
Again, I think it doesn't need fixing here+now.